File: rxpay-api/src/main/java/com/cuieney/sdk/rxpay/wechatpay/WXPayWay.java
package com.cuieney.sdk.rxpay.wechatpay;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.text.TextUtils;
import android.util.Log;

import com.cuieney.sdk.rxpay.PaymentStatus;
import com.cuieney.sdk.rxpay.RxBus;
import com.tencent.mm.opensdk.modelpay.PayReq;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

import org.json.JSONException;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;


/**
 * wechat Method of payment
 */
public class WXPayWay {
    private static final String PARTNER_ID = "partnerId";
    private static final String NONCE_STR = "nonceStr";
    private static final String TIME_STAMP = "timeStamp";
    private  static final String SIGN = "sign";
    private static final String META_WX_APPID = "WX_APPID";
    private static final String META_PARTNER_ID = "PARTNER_ID";
    private static final String META_API_KEY = "API_KEY";


    public static Flowable<PaymentStatus> payMoney(final Activity context, final JSONObject json) {

        return Flowable.create(new FlowableOnSubscribe<PaymentStatus>() {
            @Override
            public void subscribe(final FlowableEmitter<PaymentStatus> e) throws Exception {
                final String appId = getMetaData(context, META_WX_APPID);
                final IWXAPI api = WXAPIFactory.createWXAPI(context, appId);
                api.registerApp(appId);
                PayReq req = new PayReq();

                req.appId = appId;
                setValue(req,PARTNER_ID,json.optString("partnerId"),context);
                req.prepayId = json.optString("prepayId");
                setValue(req,NONCE_STR,json.optString("nonceStr"),context);
                setValue(req,TIME_STAMP,json.optString("timeStamp"),context);
                req.packageValue = json.optString("packageValue", "Sign=WXPay");
                setValue(req,SIGN,json.optString("sign"),context);
                req.extData = "app data";

                boolean sendReq = api.sendReq(req);
                if (!sendReq) {
                    e.onNext(new PaymentStatus(false));
                    e.onComplete();
                } else {
                    RxBus.getDefault().toFlowable(PaymentStatus.class)
                            .subscribe(new Consumer<PaymentStatus>() {
                                @Override
                                public void accept(PaymentStatus paymentStatus) throws Exception {
                                    e.onNext(paymentStatus);
                                    e.onComplete();
                                }
                            }, new Consumer<Throwable>() {
                                @Override
                                public void accept(Throwable throwable) throws Exception {
                                    e.onNext(new PaymentStatus(false));
                                    e.onComplete();
                                }
                            });
                }


            }
        }, BackpressureStrategy.ERROR)
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io());


    }

    public static String getMetaData(Activity context, String metaData) {
        ApplicationInfo info = null;
        try {
            info = context.getApplication().getPackageManager()
                    .getApplicationInfo(context.getPackageName(),
                            PackageManager.GET_META_DATA);
            Object data = info.metaData.get(metaData);
            if (data == null) {
                throw new NullPointerException(metaData + " field cannot be empty");
            }
            return String.valueOf(data);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }


    private static void setValue(PayReq req, String value, String rawValue, Activity context) {
        String configValue = rawValue;
        switch (value) {
            case PARTNER_ID:
                if (configValue.isEmpty()) {
                    configValue = getMetaData(context, META_PARTNER_ID);
                }
                req.partnerId = configValue;
                break;
            case NONCE_STR:
                if (configValue.isEmpty()) {
                    configValue = genNonceStr();
                }

                req.nonceStr = configValue;
                break;
            case TIME_STAMP:
                if (configValue.isEmpty()) {
                    configValue =  genTimeStamp();
                }
                req.timeStamp = configValue;
                break;
            case SIGN:
                if (configValue.isEmpty()) {
                    configValue =  genAppSign(req,getMetaData(context, META_API_KEY));
                }
                req.sign = configValue;
                break;
            default:break;
        }
    }

    private static String genNonceStr() {
        Random random = new Random();
        return getMessageDigest(String.valueOf(random.nextInt(10000)).getBytes());
    }


    private static String genAppSign(PayReq payReq, String apiKey) {
        LinkedList<NameValuePair> params = new LinkedList<>();
        params.add(new NameValuePair("appid", payReq.appId));
        params.add(new NameValuePair("noncestr", payReq.nonceStr));
        params.add(new NameValuePair("package", payReq.packageValue));
        params.add(new NameValuePair("partnerid", payReq.partnerId));
        params.add(new NameValuePair("prepayid", payReq.prepayId));
        params.add(new NameValuePair("timestamp", payReq.timeStamp));


        StringBuilder tempSb = new StringBuilder();
        for (int i = 0; i < params.size(); i++) {
            tempSb.append(params.get(i).getName());
            tempSb.append('=');
            tempSb.append(params.get(i).getValue());
            tempSb.append('&');
        }
        tempSb.append("key=");
        tempSb.append(apiKey);
        return getMessageDigest(tempSb.toString().getBytes()).toUpperCase();
    }

    private final static String getMessageDigest(byte[] buffer) {
        char hexDigits[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
        try {
            MessageDigest mdTemp = MessageDigest.getInstance("MD5");
            mdTemp.update(buffer);
            byte[] md = mdTemp.digest();
            int j = md.length;
            char str[] = new char[j * 2];
            int k = 0;
            for (int i = 0; i < j; i++) {
                byte byte0 = md[i];
                str[k++] = hexDigits[byte0 >>> 4 & 0xf];
                str[k++] = hexDigits[byte0 & 0xf];
            }
            return new String(str);
        } catch (Exception e) {
            return null;
        }
    }

    private static String genTimeStamp() {
        return String.valueOf(System.currentTimeMillis() / 1000);
    }
}

File: rxpay-api/src/main/java/com/cuieney/sdk/rxpay/wechatpay/WXPayWay.kt
package com.cuieney.sdk.rxpay.wechatpay

import android.app.Activity
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager

import com.cuieney.sdk.rxpay.PaymentStatus
import com.cuieney.sdk.rxpay.RxBus
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.openapi.WXAPIFactory

import org.json.JSONObject

import java.security.MessageDigest
import java.util.LinkedList
import java.util.Random

import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableOnSubscribe
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import java.nio.charset.Charset
import kotlin.experimental.and


/**
 * wechat Method of payment
 */
object WXPayWay {
    private val PARTNER_ID = "partnerId"
    private val NONCE_STR = "nonceStr"
    private val TIME_STAMP = "timeStamp"
    private val SIGN = "sign"
    private val META_WX_APPID = "WX_APPID"
    private val META_PARTNER_ID = "PARTNER_ID"
    private val META_API_KEY = "API_KEY"


    fun payMoney(context: Activity, orderInfo: String): Flowable<PaymentStatus> {

        return Flowable.create(FlowableOnSubscribe<PaymentStatus> { e ->
            val appId = getMetaData(context, META_WX_APPID)
            val api = WXAPIFactory.createWXAPI(context, appId)
            api.registerApp(appId)
            val req = PayReq()
            val json:JSONObject
            try {
                json = JSONObject(orderInfo)
            } catch (e: Exception) {
                throw IllegalArgumentException(e)
            }

            req.appId = appId
            val exist = setValue(req, SIGN, json.optString("sign"), context)
            if (!exist) {
                setValue(req, NONCE_STR, json.optString("null"), context)
                setValue(req, TIME_STAMP, json.optString("null"), context)
            }else{
                val nonceStrExist = setValue(req, NONCE_STR, json.optString("nonceStr"), context)
                if (!nonceStrExist) {
                    throw NullPointerException(NONCE_STR + "  FIELD CANNOT BE EMPTY")
                }
                val timeStampExist = setValue(req, TIME_STAMP, json.optString("timeStamp"), context)
                if (!timeStampExist) {
                    throw NullPointerException(TIME_STAMP + "  FIELD CANNOT BE EMPTY")
                }
            }
            
            setValue(req, PARTNER_ID, json.optString("partnerId"), context)
            req.prepayId = json.optString("prepayId")
            req.packageValue = json.optString("packageValue", "Sign=WXPay")
            req.extData = "app data"

            val sendReq = api.sendReq(req)
            if (!sendReq) {
                e.onNext(PaymentStatus(false))
                e.onComplete()
            } else {
                RxBus.default.toFlowable(PaymentStatus::class.java)
                        .subscribe({ paymentStatus ->
                            e.onNext(paymentStatus)
                            e.onComplete()
                        }, {
                            e.onNext(PaymentStatus(false))
                            e.onComplete()
                        })
            }
        }, BackpressureStrategy.ERROR)
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())


    }

    fun getMetaData(context: Activity, metaData: String): String? {
        var info: ApplicationInfo?
        try {
            info = context.application.packageManager
                    .getApplicationInfo(context.packageName,
                            PackageManager.GET_META_DATA)
            val data = info!!.metaData.get(metaData) ?: throw NullPointerException(metaData + "  FIELD CANNOT BE EMPTY")
            return data.toString()
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        return null
    }


    private fun setValue(req: PayReq, value: String, rawValue: String?, context: Activity):Boolean {
        var configValue: String? = rawValue
        var exist = true
        when (value) {
            PARTNER_ID -> {
                if (configValue!!.length <= 0 ) {
                    configValue = getMetaData(context, META_PARTNER_ID)
                    exist = false
                }
                req.partnerId = configValue
            }
            NONCE_STR -> {
                if (configValue!!.length <= 0) {
                    configValue = genNonceStr()
                    exist = false
                }

                req.nonceStr = configValue
            }
            TIME_STAMP -> {
                if (configValue!!.length <= 0) {
                    configValue = genTimeStamp()
                    exist = false
                }
                req.timeStamp = configValue
            }
            SIGN -> {
                if (configValue!!.length <= 0) {
                    configValue = genAppSign(req, getMetaData(context, META_API_KEY))
                    exist = false
                }
                req.sign = configValue
            }
            else -> {
            }
        }
        return exist
    }

    private fun genNonceStr(): String? {
        val random = Random()
        return getMessageDigest(random.nextInt(10000).toString().toByteArray(Charset.defaultCharset()))
    }


    private fun genAppSign(payReq: PayReq, apiKey: String?): String {
        var params = LinkedList<NameValuePair>()
        params.add(NameValuePair("appid", payReq.appId))
        params.add(NameValuePair("noncestr", payReq.nonceStr))
        params.add(NameValuePair("package", payReq.packageValue))
        params.add(NameValuePair("partnerid", payReq.partnerId))
        params.add(NameValuePair("prepayid", payReq.prepayId))
        params.add(NameValuePair("timestamp", payReq.timeStamp))


        val tempSb = StringBuilder()
        for (i in params) {
            tempSb.append(i.name)
            tempSb.append('=')
            tempSb.append(i.value)
            tempSb.append('&')
        }
        tempSb.append("key=")
        tempSb.append(apiKey)
        return getMessageDigest(tempSb.toString().toByteArray())!!.toUpperCase()
    }

    private fun getMessageDigest(buffer: ByteArray): String? {
        val hexDigits = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
        try {
            val mdTemp = MessageDigest.getInstance("MD5")
            mdTemp.update(buffer)
            val md = mdTemp.digest()
            val j = md.size
            val str = CharArray(j * 2)
            var k = 0
            for (i in 0 until j) {
                val byte0 = md[i]
                str[k++] = hexDigits[byte0.toInt().ushr(4) and 0xf]
                str[k++] = hexDigits[(byte0 and 0xf).toInt()]
            }
            return str.toString()
        } catch (e: Exception) {
            return null
        }

    }

    private fun genTimeStamp(): String {
        return (System.currentTimeMillis() / 1000).toString()
    }
}

